### PR TITLE
[mlir][Interfaces] Add `SubsetOpInterface` and `SubsetExtractionOpInterface`

### PR DESCRIPTION
**Pull request by matthias-springer**
There is currently an op interface for subset insertion ops (`SubsetInsertionOpInterface`), but not for subset extraction ops. This commit adds `SubsetExtractionOpInterface` to `mlir/Interfaces`, as well as a common dependent op interface: `SubsetOpInterface`.

- `SubsetOpInterface` is for ops that operate on tensor subsets. It provides interface methods to check if two subset ops operate on equivalent or disjoint subsets. Ops that implement this interface must implement either `SubsetExtractionOpInterface` or `SubsetInsertionOpInterface`.
- `SubsetExtractionOpInterface` is for ops that extract from a tensor at a subset. E.g., `tensor.extract_slice`, `tensor.gather`, `vector.transfer_read`. Current implemented only on `tensor.extract_slice`.
- `SubsetInsertionOpInterface` is for ops that insert into a destination tensor at a subset. E.g., `tensor.insert_slice`, `tensor.parallel_insert_slice`, `tensor.scatter`, `vector.transfer_write`. Currently only implemented on `tensor.insert_slice`, `tensor.parallel_insert_slice`.

Other changes:
- Rename `SubsetInsertionOpInterface.td` to `SubsetOpInterface.td`.
- Add helper functions to `ValueBoundsOpInterface.cpp` for checking whether two slices are disjoint.

The new interfaces will be utilized by a new "loop-invariant subset hoisting"
transformation. (This new transform is roughly
what `Linalg/Transforms/SubsetHoisting.cpp` is doing, but in a generic
and interface-driven way.)

### CODE REVIEW

**Review by joker-eph — COMMENTED** [review]
Seems fine to me overall, but it would be good to have another pair of eyes: @ftynse maybe?

File: mlir/include/mlir/Interfaces/SubsetOpInterface.td [hunk 1]
```diff
@@ -1,34 +1,113 @@
...
+  }];
+}
+
+def SubsetInsertionOpInterface
+    : OpInterface<"SubsetInsertionOpInterface", [SubsetOpInterface]> {
+  let description = [{
+    This interface can be implemented by ops that insert a source value into
+    a destination tensor at a specified subset. The elements in the destination
+    tensor that are overwritten by this insertion are called "subset". The
+    updates destination tensor is returned.
```
Comment (nicolasvasilache) [inline mlir/include/mlir/Interfaces/SubsetOpInterface.td:110]: nit: updated

File: mlir/include/mlir/Interfaces/SubsetOpInterface.td [hunk 2]
```diff
@@ -1,34 +1,113 @@
...
+    efficient way to describe arbitrary subsets. This interface merely provides
+    interface methods to check if two subsets are equivalent or disjoint.
+  }];
+
+  let cppNamespace = "::mlir";
+  let methods = [
+      InterfaceMethod<
+        /*desc=*/[{
+          Return "true" if this op and the given candidate subset op operate on
+          an equivalent subset. Return "false" is the two subsets are disjoint
```
Comment (nicolasvasilache) [inline mlir/include/mlir/Interfaces/SubsetOpInterface.td:47]: nit: on equivalent subsets

File: mlir/include/mlir/Interfaces/SubsetOpInterface.td [hunk 3]
```diff
@@ -1,34 +1,113 @@
...
+  ];
+
+  let verify = [{
+    return ::mlir::detail::verifySubsetOpInterface(
+        ::mlir::cast<::mlir::SubsetOpInterface>($_op));
+  }];
+}
+
+def SubsetExtractionOpInterface
+    : OpInterface<"SubsetExtractionOpInterface", [SubsetOpInterface]> {
```
Comment (nicolasvasilache) [inline mlir/include/mlir/Interfaces/SubsetOpInterface.td:77]: I had tried some of that before the existence fo dependent interfaces and it was not a great experience.
Thanks for pushing on this front!

File: mlir/include/mlir/Interfaces/SubsetOpInterface.td [hunk 4]
```diff
@@ -1,34 +1,113 @@
...
-    Ops that deal with tensor subsets come in two flavours:
-    - Insertion flavor: Ops that insert a source tensor into a destination
-      tensor at the specified subset. Such ops usually return a new destination
-      tensor and implement the `DestinationStyleOpInterface`. Insertion ops can
-      implement the `SubsetInsertionOpInterface`. Example: "tensor.insert_slice"
-    - Extraction flavor: Ops that define a tensor subset. They extract a
-      specified subset from a tensor. There is currently no op interface for
-      such ops. Example: "tensor.extract_slice"
+    This interface can be implemented by ops that operate on tensor subsets. A
+    "subset" is a part of a tensor. This interface describes the subset that
```
Comment (nicolasvasilache) [inline mlir/include/mlir/Interfaces/SubsetOpInterface.td:17]: One thing to note is that this should not be limited to tensor operands, memref operands should also be supported.

Comment (nicolasvasilache) [reply mlir/include/mlir/Interfaces/SubsetOpInterface.td:17]: we should leave this for a followup though, this PR is already properly sized

File: mlir/include/mlir/Interfaces/SubsetOpInterface.td [hunk 5]
```diff
@@ -1,34 +1,113 @@
...
+        /*retType=*/"bool",
+        /*methodName=*/"operatesOnEquivalentSubset",
+        /*args=*/(ins
+            "::mlir::SubsetOpInterface":$candidate,
+            "::llvm::function_ref<bool(Value, Value)>":$equivalenceFn)
+      >,
+      InterfaceMethod<
+        /*desc=*/[{
+          Return "true" if this op and the given candidate subset op operate on
+          disjoint subsets. Return "false" is the two subsets are equivalent,
```
Comment (nicolasvasilache) [inline mlir/include/mlir/Interfaces/SubsetOpInterface.td:59]: nit: s/is/if

File: mlir/include/mlir/Interfaces/SubsetOpInterface.td [hunk 6]
```diff
@@ -38,23 +117,25 @@ def SubsetInsertionOpInterface : OpInterface<"SubsetInsertionOpInterface"> {
...
         /*args=*/(ins)
       >,
       InterfaceMethod<
         /*desc=*/[{
-          Return the destination tensor operand.
+          Return the destination operand. The destination operand must be a
+          tensor.
+
+          This function does not have to be implemented for destination style
+          ops that exactly one "init".
```
Comment (nicolasvasilache) [inline mlir/include/mlir/Interfaces/SubsetOpInterface.td:138]: missing verb

File: mlir/lib/Dialect/Linalg/Transforms/SubsetInsertionOpInterfaceImpl.cpp [hunk 7]
```diff
@@ -9,12 +9,29 @@
...
+#include "mlir/Interfaces/SubsetOpInterface.h"
 
 using namespace mlir;
 using namespace mlir::linalg;
 
 namespace {
+struct LinalgCopyOpSubsetOpInterface
+    : public SubsetOpInterface::ExternalModel<LinalgCopyOpSubsetOpInterface,
+                                              linalg::CopyOp> {
+  bool operatesOnEquivalentSubset(
```
Comment (nicolasvasilache) [inline mlir/lib/Dialect/Linalg/Transforms/SubsetInsertionOpInterfaceImpl.cpp:21]: Can we add a comment here to explain?
I would have thought copy ops operate on equivalent subsets by construction.

**Review by nicolasvasilache — APPROVED** [review]

File: mlir/lib/Dialect/Linalg/Transforms/SubsetInsertionOpInterfaceImpl.cpp (hunk 7 above)
Comment (matthias-springer) [reply mlir/lib/Dialect/Linalg/Transforms/SubsetInsertionOpInterfaceImpl.cpp:21]: `false` is a conservative implementation. I added a better implementation for `operatesOnEquivalentSubset`. `linalg.copy` implements this interface only for bufferization purposes. It does not participate in subset hoisting, so `operatesOnEquivalentSubset` and `operatesOnDisjointSubset` are actually not needed at the moment.

**Pull request merged**